yoti-sdk-impl/src/main/java/com/yoti/api/client/spi/remote/AttributeConverter.java

```diff
@@ -0,0 +1,75 @@
+package com.yoti.api.client.spi.remote;
+
+import static com.yoti.api.client.spi.remote.call.YotiConstants.DEFAULT_CHARSET;
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.yoti.api.client.Attribute;
+import com.yoti.api.client.spi.remote.proto.AttrProto.Anchor;
+import com.yoti.api.client.spi.remote.proto.ContentTypeProto.ContentType;
+import com.yoti.api.client.spi.remote.util.AnchorCertificateUtils;
+import com.yoti.api.client.spi.remote.util.AnchorType;
+import com.yoti.api.client.spi.remote.util.AnchorCertificateUtils.AnchorVerifierSourceData;
+
+public class AttributeConverter {
+
+    private static final ObjectMapper JSON_MAPPER = new ObjectMapper();
+    private static final Logger LOG = LoggerFactory.getLogger(AttributeConverter.class);
+
+    
+    public static Attribute convertAttribute(com.yoti.api.client.spi.remote.proto.AttrProto.Attribute attribute) throws ParseException, IOException{
```

Comment:
We can add
```
import com.yoti.api.client.spi.remote.proto.AttrProto;
```

Then this class can become
```
public class AttributeConverter {

    private static final ObjectMapper JSON_MAPPER = new ObjectMapper();
    private static final Logger LOG = LoggerFactory.getLogger(AttributeConverter.class);

    public static Attribute convertAttribute(com.yoti.api.client.spi.remote.proto.AttrProto.Attribute attribute) throws ParseException, IOException {
        switch (attribute.getContentType()) {
            case STRING:
                return attributeWithSources(attribute, attribute.getValue().toString(DEFAULT_CHARSET));
            case DATE:
                return attributeWithSources(attribute, DateAttributeValue.parseFrom(attribute.getValue().toByteArray()));
            case JPEG:
                return attributeWithSources(attribute, new JpegAttributeValue(attribute.getValue().toByteArray()));
            case PNG:
                return attributeWithSources(attribute, new PngAttributeValue(attribute.getValue().toByteArray()));
            case JSON:
                return attributeWithSources(attribute, JSON_MAPPER.readValue(attribute.getValue().toString(DEFAULT_CHARSET), Map.class));
            default:
                LOG.error("Unknown type {} for attribute {}", attribute.getContentType(), attribute.getName());
                return attributeWithSources(attribute, attribute.getValue().toString(DEFAULT_CHARSET));
        }
    }

    private static Attribute attributeWithSources(AttrProto.Attribute attribute, Object value) {
        return new Attribute(attribute.getName(), value, extractMetadata(attribute, AnchorType.SOURCE), extractMetadata(attribute, AnchorType.VERIFIER));
    }

    private static Set<String> extractMetadata(AttrProto.Attribute attribute, AnchorType anchorType) {
        Set<String> entries = new HashSet<String>();
        for (Anchor anchor : attribute.getAnchorsList()) {
            AnchorVerifierSourceData anchorData = AnchorCertificateUtils.getTypesFromAnchor(anchor);
            if (anchorData.getType().equals(anchorType)) {
                entries.addAll(anchorData.getEntries());
            }
        }
        return entries;
    }

}
```

Reply:
very good one, thanks!